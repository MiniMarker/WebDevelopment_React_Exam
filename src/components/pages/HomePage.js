import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from "react-redux";
import {login} from "../../actions/auth";


export class HomePage extends React.Component {

	constructor(props) {
		super(props);
	}

	async componentDidMount() {
		await fetch("/api/user").then((res) => {
			switch(res.status) {
				case 200:

					res.json()
						.then((parsedBody) => {
							this.props.login(parsedBody.username);
						});

					break;
				case 401:
					break;
				default:
					console.log("ERROR: unexpected statuscode!", res.status);

			}
		})
	}

	renderAuthenticatedUser() {
		return (

			<div>
				<h3>Hi {this.props.auth.username}!</h3>

				<p>This is a multiplayer Quiz game where you can play against a random component online</p>

				<Link to={"/host"}>Host a game</Link>
				<Link to={"/join"}>Join a game</Link>
			</div>
		);
	}

	renderGuest() {
		return (
			<h3>No logged in user</h3>
		);
	}

	render() {
		return (
			<div className={"container"}>
				{this.props.auth.username
					? this.renderAuthenticatedUser()
					: this.renderGuest()
				}
			</div>
		);
	}
}

const mapStateToProps = (state) => {
	return {
		auth: state.auth
	};
};

const mapDispatchToProp = (dispatch) => {
	return {
		login: (username) => dispatch(login(username))
	}
};

export default connect(mapStateToProps, mapDispatchToProp)(HomePage);