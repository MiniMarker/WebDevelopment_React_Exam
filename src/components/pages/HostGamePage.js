import React from 'react';
import openSocket from "socket.io-client";
import {createTokenForUser} from "../../../server/config/tokens";

export class HostGamePage extends React.Component {

	constructor(props) {
		super(props);

		this.state = {
			name: ""
		}
	}

	async componentDidMount() {
		await fetch("/api/user").then((res) => {

			switch(res.status) {
				case 200:

					res.json()
						.then((parsedBody) => {
							this.props.login(parsedBody.username);
						});

					break;
				case 401:

					this.props.history.push("/login");

					break;
				default:
					console.log("ERROR: unexpected statuscode!", res.status);
			}
		});
	}



	onInputChange = (event) => {
		this.setState({
			[event.target.name]: event.target.value
		})
	};

	handleCreateGame = (event) => {
		event.preventDefault();

		this.props.history.push("/game");
	};

	render() {
		return (
			<div className={"container"} onSubmit={this.handleCreateGame}>
				<form className={"auth_form"}>
					<input
						name={"name"}
						type={"text"}
						onChange={this.onInputChange}
						placeholder={"Game name"}
					/>
					<button>Host game</button>
				</form>

			</div>
		);
	}
}