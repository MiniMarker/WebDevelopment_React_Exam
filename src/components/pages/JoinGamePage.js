import React from 'react';
import {Link} from 'react-router-dom';

export class JoinGamePage extends React.Component {

	constructor(props) {
		super(props);

		this.state = {
			name: ""
		}
	}

	async componentDidMount() {
		await fetch("/api/user").then((res) => {

			switch(res.status) {
				case 200:

					res.json()
						.then((parsedBody) => {
							this.props.login(parsedBody.username);
						});

					break;
				case 401:

					this.props.history.push("/login");

					break;
				default:
					console.log("ERROR: unexpected statuscode!", res.status);
			}
		});
	}

	onInputChange = (event) => {
		this.setState({
			[event.target.name]: event.target.value
		})
	};

	handleCreateGame = (event) => {
		event.preventDefault();

		this.props.history.push("/game");
	};

	render() {
		return (
			<div className={"container"} onSubmit={this.handleCreateGame}>
				<li>
					<ul><Link to={"/game"}>Game 1</Link></ul>
					<ul><Link to={"/game"}>Game 2</Link></ul>
					<ul><Link to={"/game"}>Game 3</Link></ul>
				</li>
			</div>
		);
	}


}