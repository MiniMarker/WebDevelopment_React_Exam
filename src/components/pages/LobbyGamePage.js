import React from 'react';
import openSocket from 'socket.io-client';
import {connect} from "react-redux";
import {login, logout} from "../../actions/auth";
import {Quiz} from "./Quiz";

class LobbyGamePage extends React.Component {

	constructor(props) {
		super(props);

		this.state = this.setDefaultState();
	}

	setDefaultState = () => {

		console.log("State reset to default values");

		return {
			game: null,
			opponentIds: [],
			numberOfPlayers: null,
			errorMsg: null
		}
	};

	// ############## LIFECYCLE FUNCTIONS ##############
	componentDidMount = () => {
		//join or start game

		/*
		*  In this component i don't handle page refreshes for keeping frontend login state
		* */

		const username = this.props.auth.username;

		if(username === null) {
			this.setState({errorMsg: "You have to login before entering a game"});
			this.props.history.push("/login");
			return;
		}

		this.setupWsSockets();

		this.doLoginWebSocket().then(
			this.joinGame
		);
	};

	componentWillUnmount() {

		if(this.state.game !== null) {
			this.socket.emit("userLeftGame", (this.state.game.id));
		}

		this.socket.disconnect();
	}


	setupWsSockets = () => {
		// Open the socket
		this.socket = openSocket(window.location.origin);


		//Subscribe to emits from "update"
		this.socket.on("update", (dto) => {

			if(dto === null || dto === undefined) {
				this.setState({errorMsg: "Invalid response from server."});
				return;
			}

			if(dto.error !== null && dto.error !== undefined) {
				this.setState({errorMsg: dto.error});
				return;
			}

			console.log(dto.data);
		});

		this.socket.on("renderGame", (data) => {

			this.setState({
				game: data.game,
				errorMsg: data.errorMsg
			});
		});

		this.socket.on("updateNumberOfPlayers", (data) => {
			this.setState({
				numberOfPlayers: data
			})
		});

		//Subscribe to emits from "disconnect"
		this.socket.on("disconnect", () => {
			this.setState({errorMsg: "Disconnected from Server."});
		});

		//Subscribe to emits from "login"
		this.socket.on("login", () => {
			this.setState({errorMsg: "Connected to Server."});
		});
	};

	doLoginWebSocket = async() => {

		let response = await fetch("/api/wstoken", {method: "post"});

		await response.json()
			.then((data) => {
				this.socket.emit("login", data);
			})
			.catch((e) => {
				console.error("Error on parse to JSON", e);
			});
	};


	joinGame = () => {
		let data = {
			username: this.props.auth.username
		};

		this.socket.emit("joinGame", (data));
	};


	startGame = () => {

		this.socket.emit("startGame");
	};

	// ############## RENDER FUNCTIONS ##############
	renderIdleText = () => (
		<div>
			<h3>Wait for people to join!</h3>

			{this.state.numberOfPlayers
				? <h3>{this.state.numberOfPlayers} people has joined the game</h3>
				: <h3>x people has joined the game</h3>}

			<img src={"images/loader.gif"}/>

			<br/>

			<button onClick={this.startGame}>Start game!</button>
		</div>
	);

	renderGuest = () => (
		<h3>Why are you even seeing this?! you should've been redirected to login by now........</h3>
	);

	render() {
		return (
			<div className={"container"}>

				<p>{this.state.errorMsg}</p>
				<br/><br/>

				{this.props.auth.username
					? this.state.game
						? <Quiz
							game={this.state.game}
							socket={this.socket}
							authUser={this.props.auth.username}
						/>
						: this.renderIdleText()
					: this.renderGuest()
				}
			</div>
		);
	}
}

// ############## REDUX FUNCTIONS ##############
const mapStateToProps = (state) => {
	return {
		auth: state.auth
	};
};

const mapDispatchToProp = (dispatch) => {
	return {
		login: (username) => dispatch(login(username)),
		logout: () => dispatch(logout())
	}
};

export default connect(mapStateToProps, mapDispatchToProp)(LobbyGamePage);