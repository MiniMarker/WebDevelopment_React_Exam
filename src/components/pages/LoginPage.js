import React from 'react';
import {connect} from "react-redux";
import {login} from "../../actions/auth";

export class LoginPage extends React.Component {

	constructor(props) {
		super(props);

		this.state = {
			username: "",
			password: "",
			errorMsg: null
		};
	};

	onUsernameChange = (event) => {
		this.setState({
			username: event.target.value
		})
	};

	onPasswordChange = (event) => {
		this.setState({
			password: event.target.value
		})
	};

	handleLogin = async (event) => {

		event.preventDefault();

		const {username, password} = this.state;
		const url = "/api/login";
		const payload = {username: username, password: password};

		let response;

		try {
			response = await fetch(url, {
				method: "post",
				headers: {
					"Content-Type": "application/json"
				},
				body: JSON.stringify(payload)
			});
		} catch (err) {
			this.setState({ errorMsg: "Failed to connect to server: " + err });
			return;
		}

		if (response.status === 401) {
			this.setState({ errorMsg: "Invalid userId/password" });
			return;
		}

		if (response.status !== 204) {
			this.setState({
				errorMsg:
					"Error when connecting to server: status code " + response.status
			});
			return;
		}

		this.props.login(username);
		this.setState({ errorMsg: null });
		this.props.history.push("/");

	};


	render() {
		return (
			<div className={"container"}>
				<form className={"auth_form"} onSubmit={this.handleLogin}>
					<input
						type={"text"}
						onChange={this.onUsernameChange}
						placeholder={"Username"}
					/>
					<input
						type={"text"}
						onChange={this.onPasswordChange}
						placeholder={"Password"}
					/>
					<button>Login</button>
				</form>
			</div>
		);
	}
}

const mapDispatchToProp = (dispatch) => {
	return {
		login: (username) => dispatch(login(username))
	}
};

export default connect(undefined, mapDispatchToProp)(LoginPage);