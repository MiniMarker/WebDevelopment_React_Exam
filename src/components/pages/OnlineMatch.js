import React from "react";
import openSocket from "socket.io-client";
import {login} from "../../actions/auth";
import connect from "react-redux/es/connect/connect";
import {Link} from "react-router-dom";

class OnlineMatch extends React.Component {
	constructor(props) {
		super(props);
	}

	async componentDidMount() {
		await fetch("/api/user").then((res) => {

			switch(res.status) {
				case 200:

					res.json()
						.then((parsedBody) => {
							this.props.login(parsedBody.username);

							this.socket = openSocket(window.location.origin)

						});

					break;
				case 401:

					this.props.history.push("/login");

					break;
				default:
					console.log("ERROR: unexpected statuscode!", res.status);
			}
		});
	}

	static renderAuthenticatedUser() {
		return (
			<div>
				<h3>Wait for people to join!</h3>

				<img src={"images/loader.gif"}/>
			</div>
		);
	}

	static renderGuest() {
		return (
			<h3>Why are you even seeing this?! you should've been redirected to login by now........</h3>
		);
	}

	render() {
		return (
			<div className={"container"}>
				{this.props.auth.username
					? OnlineMatch.renderAuthenticatedUser()
					: OnlineMatch.renderGuest()
				}
			</div>
		);
	}
}

const mapStateToProps = (state) => {
	return {
		auth: state.auth
	};
};

const mapDispatchToProp = (dispatch) => {
	return {
		login: (username) => dispatch(login(username))
	}
};


export default connect(mapStateToProps, mapDispatchToProp)(OnlineMatch);