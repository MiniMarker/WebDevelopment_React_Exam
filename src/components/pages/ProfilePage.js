import React from 'react'
import {connect} from "react-redux";
import {logout} from "../../actions/auth";

class ProfilePage extends React.Component {
	constructor(props) {
		super(props);
	}

	handleLogout = async () => {
		const url = "/api/logout";

		let response;

		try {
			response = await fetch(url, { method: "post" });
		} catch (err) {
			alert("Failed to connect to server: " + err);
			return;
		}

		if (response.status !== 204) {
			alert("Error when connecting to server: status code " + response.status);
			return;
		}

		console.log("Logout success");

		this.props.logout();
		this.props.history.push("/");
	};

	render() {
		return (
			<div>
				<h1>ProfilePage</h1>
				<button onClick={this.handleLogout}>Log out</button>
			</div>
		);
	}
}

const mapDispatchToProp = (dispatch) => {
	return {
		logout: () => dispatch(logout())
	}
};


export default connect(undefined, mapDispatchToProp)(ProfilePage);