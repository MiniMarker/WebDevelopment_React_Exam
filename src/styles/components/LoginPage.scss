.loginBtn{
	height: 30px;
	cursor: pointer;
	background-color: red;
	width: 50px;
}