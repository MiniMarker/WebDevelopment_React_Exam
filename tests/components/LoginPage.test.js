import React from 'react';
import { shallow } from 'enzyme';
import {LoginPage} from "../../src/components/pages/LoginPage";

let loginSpy, wrapper;

beforeEach(() => {

	loginSpy = jest.fn();

	wrapper = shallow(<LoginPage login={loginSpy}/>)

});

test("RenderLoginPage", () => {
    expect(wrapper).toMatchSnapshot();
});
/*
test("CallLoginOnLoginButtonClick", () => {
    wrapper.find("input").at(0).simulate("change", { target: { value: "username" }});
	wrapper.find("input").at(1).simulate("change", { target: { value: "password" }});
	wrapper.find("button").simulate("click");

	expect(loginSpy).toHaveBeenCalled();

});
*/