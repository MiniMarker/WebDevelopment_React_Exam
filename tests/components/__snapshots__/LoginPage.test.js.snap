// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderLoginPage 1`] = `
<div
  className="container"
>
  <form
    className="auth_form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Username"
      type="text"
    />
    <input
      onChange={[Function]}
      placeholder="Password"
      type="text"
    />
    <button>
      Login
    </button>
  </form>
</div>
`;
